src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -168,29 +196,88 @@ struct TensorContract
+  return std::pair{std::numeric_limits<size_t>::max(),
+                   std::numeric_limits<size_t>::max()};
 }
 
-namespace detail {
-// Helper struct to allow contractions by using repeated indices in operator()
-// calls to tensor.
-template <template <typename> class TE, typename ReplacedArgList, size_t I,
-          typename TotalContracted, typename T>
-SPECTRE_ALWAYS_INLINE static constexpr auto fully_contract(const T& t) {
-  if constexpr (I == 2 * (TotalContracted::value - 1)) {
-    return contract<ti_contracted_t<I>::value, ti_contracted_t<I + 1>::value>(
-        TE<ReplacedArgList>(t));
+/*!
+ * \ingroup TensorExpressionsGroup
+ * \brief Creates a contraction expression from a tensor expression if there are
+ * any indices to contract
+ *
+ * \details If there are no indices to contract, the input TensorExpression is
+ * simply returned. If not, a contraction expression is created for contracting
+ * one pair of upper and lower indices. If there is more than one pair of
+ * indices to contract, subsequent contraction expressions are recursively
+ * created, nesting one contraction expression inside another.
+ *
+ * For example, if we have tensor \f${R^{ab}}_{ab}\f$ represented by the tensor
+ * expression, `R(ti_A, ti_B, ti_a, ti_b)`, then one contraction expression is
+ * created to represent contracting \f${R^{ab}}_ab\f$ to \f${R^b}_b\f$, and a
+ * second to represent contracting \f${R^b}_b\f$ to the scalar, \f${R}\f$.
+ *
+ * @param t the TensorExpression to potentially contract
+ * @return the input tensor expression or a contraction expression of the input
+ */
+template <typename T, typename X, typename Symm, typename IndexList,
+          typename... TensorIndices>
+SPECTRE_ALWAYS_INLINE static constexpr auto contract(
+    const TensorExpression<T, X, Symm, IndexList, tmpl::list<TensorIndices...>>&
+        t) noexcept {
+  constexpr std::array<size_t, sizeof...(TensorIndices)> tensorindex_values = {
+      {TensorIndices::value...}};
+  constexpr std::pair first_tensorindex_values_to_contract =
+      get_first_tensorindex_values_to_contract(tensorindex_values);
+  constexpr std::pair no_indices_to_contract_sentinel{
+      std::numeric_limits<size_t>::max(), std::numeric_limits<size_t>::max()};
+
+  if constexpr (first_tensorindex_values_to_contract ==
+                no_indices_to_contract_sentinel) {
+    // There aren't any indices to contract, so we just return the input
+    return ~t;
   } else {
-    return contract<ti_contracted_t<I>::value, ti_contracted_t<I + 1>::value>(
-        fully_contract<TE, ReplacedArgList, I + 2, TotalContracted>(t));
+    // We have a pair of indices to be contract
+    using tensorindex_to_contract =
+        TensorIndex<first_tensorindex_values_to_contract.first>;
+    using opposite_tensorindex_to_contract =
+        TensorIndex<first_tensorindex_values_to_contract.second>;
+    return contract(TensorContract<tensorindex_to_contract,
```

Comment:
[optional] I'm now wondering if it makes more sense to pass in the locations of the indices to contract, rather the indices them selves. Combined with guaranteeing `upper, lower` in the ordering I think it also means you don't need to do the `tmpl::index_of` call inside the `TensorContract` struct. It might then be good to move `TensorContract` into the `detail` namespace, making this function the interface to it. Thoughts?

Reply:
I don't have a strong preference on this. I think I originally had it this way with the `TensorIndex`s instead of their positions just so that `TensorContract` could work out details internally in one place instead of it being externally dictated in multiple. If there were multiple spots where `TensorContract` would be called, then we might have to work out the index positions each time, which seemed like more work to use `TensorContract` as an interface than just giving it the two `TensorIndex`s. It was like it kind of felt more fragile for something external to be telling `TensorContract` the positions of the two indices instead of `TensorContract` working it out for itself, especially with both a contracted args list and uncontracted args list as things to deal with. However, since it seems `TensorContract` shouldn't really be used anywhere else, I don't think this point matters so much. I'll make this change!

Regarding putting `TensorContract` in the `detail` namespace, I think I don't naturally see it there only because it's one of the three expression types making up this tensor expression library/interface. So it doesn't really feel like background details, as a whole. Is there a property about it that makes you feel it belongs in the `detail` namespace, whereas `AddSub` and `Product` wouldn't? I think I'm not as familiar with when a whole class/struct makes more sense to stick in `detail`.

Reply:
That all sounds good. Let's not move it into a detail namespace. I agree consistency with `AddSub` and `Product` will be good!

Reply:
Done! (Except, instead of the upper valence index's position and lower valence index's position, I return the smaller position and lower position, regardless of valence)